worlds/sm64ex/Rules.py

```diff
@@ -1,33 +1,54 @@
-from ..generic.Rules import add_rule
-from .Regions import connect_regions, sm64_level_to_paintings, sm64_paintings_to_level, sm64_level_to_secrets, sm64_entrances_to_level, sm64_level_to_entrances
+from typing import Callable, Union
 
-def shuffle_dict_keys(world, obj: dict) -> dict:
-    keys = list(obj.keys())
-    values = list(obj.values())
+from BaseClasses import MultiWorld
+from ..generic.Rules import add_rule, set_rule
+from .Locations import location_table
+from .Regions import connect_regions, sm64_level_to_paintings, sm64_paintings_to_level, sm64_level_to_secrets, \
+sm64_level_to_secrets, sm64_entrances_to_level, sm64_level_to_entrances
+from .Items import action_item_table
+
+def shuffle_dict_keys(world, dictionary: dict) -> dict:
+    keys = list(dictionary.keys())
+    values = list(dictionary.values())
     world.random.shuffle(keys)
-    return dict(zip(keys,values))
+    return dict(zip(keys, values))
 
-def fix_reg(entrance_ids, entrance, destination, swapdict, world):
-    if entrance_ids[entrance] == destination: # Unlucky :C
+def fix_reg(entrance_map, entrance, destination, swapdict, world):
+    if entrance_map[swapdict[entrance]] == destination: # Unlucky :C
         rand = world.random.choice(swapdict.keys())
-        entrance_ids[entrance], entrance_ids[swapdict[rand]] = rand, entrance_ids[entrance]
-        swapdict[rand] = entrance_ids[entrance]
+        entrance_map[entrance], entrance_map[swapdict[rand]] = rand, entrance_map[entrance]
+        swapdict[rand] = entrance_map[entrance]
         swapdict.pop(entrance)
 
-def set_rules(world, player: int, area_connections: dict):
+def set_rules(world, player: int, area_connections, star_costs, move_rando_bitvec: int):
     randomized_level_to_paintings = sm64_level_to_paintings.copy()
     randomized_level_to_secrets = sm64_level_to_secrets.copy()
-    if world.AreaRandomizer[player].value == 1:  # Some randomization is happening, randomize Courses
+    valid_move_randomizer_start_courses = [
+        "Bob-omb Battlefield", "Jolly Roger Bay", "Cool, Cool, Mountain",
+        "Big Boo's Haunt", "Lethal Lava Land", "Shifting Sand Land",
+        "Dire, Dire Docks", "Snowman's Land"
+    ]  # Excluding WF, HMC, WDW, TTM, THI, TTC, and RR
+    if world.AreaRandomizer[player].value >= 1:  # Some randomization is happening, randomize Courses
         randomized_level_to_paintings = shuffle_dict_keys(world,sm64_level_to_paintings)
+        if world.AreaRandomizer[player].value < 3 and move_rando_bitvec > 0:
+            first_course = world.random.choice(valid_move_randomizer_start_courses)
+            for entrance, painting in randomized_level_to_paintings.items():
```

Comment:
I dont like this loop.
Equivalent but to me nicer looking: `next(key for key, value in dd.items() if value == 'value')
`.
Though tbh I found this on stackoverflow

Reply:
I normally use comprehensions everywhere, I'm not sure why I didn't do so here.  Changed.  (https://github.com/ArchipelagoMW/Archipelago/pull/2569/commits/29480bd177069995797952def1dd5390dcb651f0)